Adding Jitamin support

gumulka opened the pull request:
I'm not sure, if this should be integrated upstream, as Jitamin is quite old, unmaintained and has no API, so this Plugin relies on web-scraping to get the needed information. 

But it is the task-management solution my employer uses, and I use taskwarrior, so I wrote this.

gumulka committed:
Added first draft for Jitamin service

This is a somewhat working version

Importing task is working, the fields also all seem to be
working without problems.

ryneeverett commented on bugwarrior/services/jitamin.py:
```diff
@@ -0,0 +1,240 @@
+
+class JitaminConfig(config.ServiceConfig, prefix="jitamin"):
+    service: typing_extensions.Literal["jitamin"]
+    base_uri: config.StrippedTrailingSlashUrl
```
This type exists for backwards compatibility. I'd prefer to use one of the [upstream types](https://pydantic-docs.helpmanual.io/usage/types/#urls) such as `pydantic.AnyUrl` if possible.

ryneeverett commented on bugwarrior/services/jitamin.py:
```diff
@@ -0,0 +1,240 @@
+    URL = "jitaminurl"
+    POSITION = "jitaminposition"
+    COLUMN = "jitamincolumn"
+    DESCRIPTION = "jitamindescription"
```
Description is a core attribute.[^1] Is there a reason we need a separate UDA for this?

[^1]: https://taskwarrior.org/docs/udas.html

gumulka replied:
Jitamin task consists of a short title and a long description. Kind of like Gitlab issues. I use the title for the UDA description and create my own UDA description for the often long multiline description.

I could opt to make it an annotation, but they also do not feel so good to me. Open to discuss here.

ryneeverett commented on bugwarrior/docs/services/jitamin.rst:
```diff
@@ -0,0 +1,55 @@
+=======
+
+You can import tasks from your Jitamin instance using
+the ``jitamin`` service name.
```
This would be a good place to note how to install the extra dependencies.

ryneeverett commented on bugwarrior/services/jitamin.py:
```diff
@@ -0,0 +1,240 @@
+class JitaminIssue(Issue):
+    URL = "jitaminurl"
+    POSITION = "jitaminposition"
+    COLUMN = "jitamincolumn"
```
It seems a little odd to have column as a separate UDA and also optionally use it as a tag. Maybe it makes sense to Jitamin users though.

gumulka replied:
No, it was more of a personal preference. I like to see the column in my default task view, which does not include the UDA, so tag seemed to be the first logical value.

ryneeverett commented on bugwarrior/services/jitamin.py:
```diff
@@ -0,0 +1,240 @@
+        while True:
+            tasklist = self.get_task_list_page(
+                "?controller=SearchController&action=index&q="
+                + self.config.query + " status:open"
```
We typically implement the `query` option to be as open ended as possible, which usually means it would be an alternative to the default query string rather than added to it. For example, it appears that one would not be able to query for closed issues the way this is implemented.

gumulka replied:
I first had it as an open query, but then seen, that closed issues are reopend by bugwarrior per default (see https://github.com/ralphbean/bugwarrior/blob/develop/bugwarrior/db.py#L380), only to then be closed by taskwarrior, as they have an ending, which lead to a lot of lines like this:

```
INFO:bugwarrior.db:Updating task c9de8aa8-c329-47f7-89b3-540396fdb2cc, (bw)# - root Passwort not changable vial co; status: 'completed' -> 'pending'; end: datetime.datetime(2022, 2, 25, 12, 34, tzinfo=tzutc()) -> datetime.datetime(2022, 2, 25, 12, 34, tzinfo=<UTC>)
```
every time I do `bugwarrior pull`

So instead of filling the end field, I decided to skip all closed tasks.

ryneeverett reviewed (commented):
I don't have a problem with merging support for this service if people are still using it.

Aside from the inline comments, I'd like to see at least a minimal test module before to merging this. It may help to pull out the requests and beautiful soup logic into a separate "client" class. This isn't a dealbreaker but it provides separation of concerns and might make mocking easier.

gumulka committed:
Using pydantic type instead of legacy

gumulka committed:
Moved jitamin handling to new class

gumulka commented:
Hey, I "moved" the Jitamin logic to a new class. Will move the new class to a new project later on, but just wanted some feedback if that is what you had in mind.

Will add tests later this week.

gumulka committed:
Stop logging in on init, move to page request

Only logging on on page request, lets us better mock the Jitamin class
as there are no request during init.

ryneeverett commented:
> Hey, I "moved" the Jitamin logic to a new class. Will move the new class to a new project later on, but just wanted some feedback if that is what you had in mind.

Yes, the way the client class abstracts away the fetching/parsing from the service class is exactly what I had in mind.

I understand why you added the lazy `Task` class but I think it's somewhat confusing given the conceptual overlap with the `Issue` class. We don't have anything like this in any other bugwarrior service so there's no precedent, but I think the terminology might match up better if `Task` was renamed `Record` since that aligns with the fact that it represents remote data prior to the conversion into local taskwarrior data (akin to `Issue.record`). What do you think?

gumulka commented:
The week after I created the pull request. My employer decided to ditch Jitamin and move to Jira. Since this is now a permanent decision and I won't be using this anymore and will not be able to fix any bugs, I'm closing this merge request.
